### DISCUSSION TITLE
DeepSeek-V3.2 Decode Perf Analysis

### DISCUSSION THREAD

**Discussion by mvasiljevicTT** [tenstorrent/tt-mlir#8504] [Ideas]
## Model bringup

Currently, DeepSeek-V3.2 is brought up only for the **2-layer** benchmark configuration (not full model yet).

Current PCC on this setup:
- **Prefill PCC verification passed with PCC = 0.929178**
- **First decode PCC verification passed with PCC = 0.932751**

Important behavior we also still see: different users produce different outputs

Full model has 61 layers - first 3 dense and then 58 MoE. Loaded with `num-layers=1` it has just one MoE layer, `num-layers=2` has one dense and then one MoE.

Based on device time for those two settings, total perf can be estimated as `T_layers_61 ~= 3*(T2 - T1) + 58*T1 = 3*T2 + 55*T1`.

Each layer is like:
```
input
  → RMSNorm → Attention sub-layer → residual add
  → RMSNorm → FFN sub-layer (dense or MoE) → residual add
output
```
so attention and MoE blocks can be analyzed separately, like we did for gpt

## Device perf

On wh galaxy in nightly CI shows 500-570ms per iteration for 2lyrs (high variance is expected for MoE models in general, not sure yet for this implementation)

Locally on bh galaxy it is ~450 per iteration (will update with actual numbers later)

Getting tracy the usual way
```
tracy -p -r --sync-host-device -m \
  pytest -svv tests/benchmark/test_llms.py::test_deepseek_v3_2_exp_tp_galaxy_2_layers \
  --decode-only --num-layers 2 --max-output-tokens 3 
```
gets randomly big one `AllToAllCombineDeviceOperation` of 85ms and one `ReduceScatterDeviceOperation` of 72ms, leading to total device time of 470ms, 56% of which is SparseMatmulDeviceOperation (267ms). 

<img width="2804" height="500" alt="image" src="https://github.com/user-attachments/assets/19c98d44-85de-475c-868e-b88367270942" />

In other iterations those RS and AtAC ops randomly work ok, which is expected because with this device perf we could never achieve 450ms e2e.

Generating and using py code from this branch [benchmark -> codegen](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/codegen-py-export-hook):
```
pytest -svv tests/benchmark/test_llms.py::test_deepseek_v3_2_exp_tp_galaxy_2_layers \
  --decode-only --num-layers 2 --max-output-tokens 3 \
  --codegen-py-export-path deepseek_codegen

run -t deepseek_codegen/graph_0
```
gives as more realistic device perf (not saying this from any theoretical point, just in this case it showed that way) with RS and AtAC being reasonably fast

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/4049a7a0-53b4-4370-bb22-764c07476793" />

Sparse matmuls are massive bottleneck, fusing them into `deepseek_moe_post_combine_tilize` is expected to solve this.

Assuming device perf of ~320ms per iteration for 2lyr, we expect ~19s per iteration of full model. 

cc: @odjuricicTT

**Comment by mvasiljevicTT** [comment]
## EmitPy iterations

Using skills from [this branch](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/tuning-skills), loosely inspired by autoresearcher, we get the perf of 80,6ms on emitted model:

Initial device perf             |  Initial autotune
:-------------------------:|:-------------------------:
<img width="600" alt="summary_emitpy_baseline" src="https://github.com/user-attachments/assets/281495ae-9b62-4dba-bf15-3832b24b27c1" /> | <img width="600" alt="summary" src="https://github.com/user-attachments/assets/039208ff-5b74-4ed3-bce1-24967a3a52e7" /> 

Full progress is on [this branch](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/deepseek-emitpy-tuning).

At each step it updates its [tuning log](https://github.com/tenstorrent/tt-xla/blob/mvasiljevic/deepseek-emitpy-tuning/deepseek_codegen/TUNING_LOG.md). 

---
**TL;DR — DeepSeek-V3.2 decode: 320k → 81k μs (−74.8%) across 37 experiments on `mvasiljevic/deepseek-emitpy-tuning`.**

  - **Where it came from:** one structural MoE fusion (E30 stacked gate+up sparse_matmul, −23%) plus its knob sweep (E31–E34 
  grid/per_core_N/in0_block_w, −51% more), and two semantic-redundancy scatter-coalesces (E8/E9, −3%). Everything else was sub-percent noise.
  - **Lesson that paid:** `tt-perf-report`'s "DRAM-bound / 2 GB/s" hint is *per-core* — the matmul looked bandwidth-bound but was actually 
  core-bound at 21/110 cores. Sizing the grid to the work (not the device max) unlocked the cascade.
  - **Lesson that didn't:** fancy fused ops (`rms_norm_post_all_gather`, `moe_compute`, QKV-heads) mostly didn't fit — WH-TG-only kernels,
  rank-4 slice assertions, or wrong tensor packing for DeepSeek MLA's latent split.
  - **Converged at E34.** Sparse_matmul knobs exhausted (E35 flatlined, E36/E37 hung); MLA QKV fusion infeasible; remaining lanes (L1-sharded 
  input, Tilize audit, RMSNorm slice.cpp workaround) are multi-iteration projects needing a different cadence.

---

I tried nudging it towards complex fusings, without specifying exact ones, it found and investigated deepseek specific ones but came to the conclusion that they can't be used here either because it is blackhole, or because of the sharding. Didn't fact check this yet.

These skills could be modified for the autoresearcher, but I thought I would have better control in nudging it in different direction to get a sense of how good is it in each. Now that even this way it failed to do any complex deepseek fusing, I have to check if that is really impossible, impossible for the agent (which I doubt), or just wrongly prompted.

For most steps it did a lot of thinking, looking through metal op implementations, and ended up doing some one line change. Didn't see any change like `move this permute 5 ops down`, but it had some interesting ideas, making additional scripts to analyze and modify the `main.py`, like for solving the scatter ops.

It wasn't specified in the skill what to do if perf is marginally improved, sometimes it kept and sometimes it reverted the change. We will need additional check there on *does it make the code better* even though it doesn't affect perf. 

All this was done on full 2 lyr model. It would be better if it was first sliced into blocks, like attention and MoE, and then analyzed and optimized separately.

**Reply by mvasiljevicTT** [reply]
With more direct prompting agent takes more time on each iteration instead of looking for easy wins, and was able to make some more complex fusings on [this branch](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/deepseek-router-fuse/deepseek_codegen) 

----

**TLDR — `mvasiljevic/deepseek-router-fuse` branch (14 commits since `mvasiljevic/deepseek-emitpy-tuning`):**

  | # | What | Δ per-step | Note |
  |---|---|---:|---|
  | E38 | Router block → `ttnn.experimental.deepseek_grouped_gate` | −25,937 μs | First big win on the branch. Second-order effect cleaned sparse_matmul sparsity. |
  | (survey) | Catalogue of all `experimental/deepseek_*` ops — most parked as prefill-only or topology-incompatible | — | Documented in `850169f76`. |
  | E39 (attempt) | `flash_multi_latent_attention_decode` first try | — | Crashed; later traced to cold-compile timeout, not a real hang. |
  | E40 + E41 | MoE host round-trips collapsed + MLA SDPA fused (both layers) | −4,220 μs | Biggest single win after E38. SDPA + cur_pos + YARN scale. |
  | E42 | Delete the now-dead MLA chains (post-E41 orphans) | −3,701 μs | Clean win, no math change. |
  | (signposts) | Added 11 tracy signposts (`prologue / attn_0 / mlp_0 / attn_1 / moe / lm_head`) + per-region perf + `PER_REGION_ANALYSIS.md` | — | Sum of regions = full `decode_1`, no gap. |
  | E43 | Transitive AST-based DCE on the now-dead V3.2 indexer compute (also reverted a broken LM-head argmax patch) | −1,767 μs net | 839 LoC removed; biggest auto-DCE pass on the branch. |
  | E44 (attempt) | TM-reorder around `sparse_matmul_2` | +803 μs | Reverted. Lesson: RM-permute / shape-dependent Untilize aren't free. |
  | E45 | Fold MoE routing-weight chain FP32 → BF16 | −35 μs | Small but clean dtype fold (matmul_29 + drop two typecasts). |
  | (exit) | Exit notes + per-region warm artifacts for E42 and E45 | — | Remaining levers (RoPE swap, async CCL, paged-cache reformat) parked as separate projects. |

  **Cumulative (2-layer test graph)**: per-step decode **80,602 → 24,038 μs (−70 %)** on this branch, **~298k → 24,038 μs (−92 %)** vs the parity-restored E0 baseline. Golden PCC held at **0.92** throughout.

  **Full-model extrapolation (DeepSeek V3.2, 61 layers = 3 dense + 58 MoE)** using post-E45 per-region constants (`pre=20 μs`, `attn=2,702.5 μs`, `mlp_dense=649 μs`, `moe=15,519 μs`, `lm_head=2,445 μs`):

  full_decode = 20 + 61 × 2,702.5 + 3 × 649 + 58 × 15,519 + 2,445
              ≈ 1,069,366 μs
              ≈ 1.069 s / decode token
              ≈ 0.935 tok/s

  (vs ≈ 0.87 tok/s at E42 — the E43 DCE win on attention + prologue translates to ~+7 % at full-model scale. MoE remains ~85 % of the full-model decode budget.)

----

Attention block             |  MoE block
:-------------------------:|:-------------------------:
<img width="400" alt="E45warm_attn_1_summary" src="https://github.com/user-attachments/assets/730a2a10-a004-46f7-b15c-3f17eb1bdca5" /> | <img width="400"  alt="E45warm_moe_summary" src="https://github.com/user-attachments/assets/975a6a76-cd57-49d5-b75d-37bd001a1d43" />

RoPE fusing is expected to solve some of the attention permutes, plan is on the branch stating that it will need more iterations to make it work. Doing that next.

**Comment by mvasiljevicTT** [comment]
After 47 tuning iterations on the EmitPy-generated decode graph, device time landed at **~22.5k μs (≈ −92% from baseline; projected ~1.02 s/token, ~0.98 samples/sec on the full 61-layer model)** — mostly from one structural sparse_matmul fusion + its knob sweep, two experimental kernel swaps (router + SDPA), and DCE of the chains they displaced.

  Three docs on the branch summarising where this landed:

  - [TUNING_SUMMARY.md](https://github.com/tenstorrent/tt-xla/blob/mvasiljevic/deepseek-router-fuse/deepseek_codegen/TUNING_SUMMARY.md) — what to do if redoing from scratch.
  - [FUSION_NOTES.md](https://github.com/tenstorrent/tt-xla/blob/mvasiljevic/deepseek-router-fuse/deepseek_codegen/FUSION_NOTES.md) — fusions landed / blocked / never-tried, with verified reasons.
  - [TT_METAL_ISSUES.md](https://github.com/tenstorrent/tt-xla/blob/mvasiljevic/deepseek-router-fuse/deepseek_codegen/TT_METAL_ISSUES.md) — tt-metal issues behind the blockers, triaged against existing ones.

  cc: @odjuricicTT

**Comment by mvasiljevicTT** [comment]
## Update 

**Now: ~0.15 s/token · ~6.5 samples/sec — ~6.7× since E47** (E47 = 1.02 s/token, 0.98 samp/s).
**Correctness:** next-token argmax = 100% throughout; each effort ≈-numerically-equal to the prior step. 
Golden-logits PCC fell ~0.92 → **0.90** at the bf4 `moe_compute` swap (effort 1) and held there (bf4 floor).

Hand-tuned the EmitPy [`main.py`](https://github.com/tenstorrent/tt-metal/blob/mvasiljevic/deepseek-decode-sharding/deepseek_codegen/graph_0/main.py) using [ttnn tuning skill](https://github.com/tenstorrent/tt-xla/blob/f17b5039d12ffc3a529a8cbddf3dfe197b6c741f/.claude/skills/tune-emitpy-perf/SKILL.md)

**Device perf** (per decode step; 61 layers = 3 dense + 58 MoE):
```
prologue +  attn×61 + dense×3 +   moe×58 + lm_head
   0.019 + 0.890×61 + 0.619×3 + 1.739×58 +   2.439 ≈ 158.5 ms
```

| effort | what | commits | speedup |
|---|---|---|---|
| **Fused MoE experts** | sparse_matmul path → `ttnn.experimental.moe_compute` (bf4) | `dcec039`,`6bca510` | **+314%** |
| **Attention layout** | MLA chains→L1, ~~`permute→ttnn.transpose`~~, reshape elim, RoPE op | `bf666df` | +25% |
| **MoE tail precision** | BF16 weighted-k-sum → `FastReduceNC` fast path (170→15 µs) | `012ba4c`,`70d3616` | +15% |
| **Reshard + fuse collectives** | local `mesh_partition` reshard → qkv-down 3 reductions→1 `all_gather+sum`; router too | `16b3ab2`,`f8cebd8`,`b3e0f41` | +7% |

**Blocked (biggest lever):** MoE = ~59% of device perf, but expert/combine restructuring trips the `moe_compute selective_reduce_combine` L1-overlap — **#46208** (barrier #43444 fixed by #45764; residual `num_buffers 14→13`).

**Compiler could auto-emit** (each generalizes a manual win above):
- ~~lower transposing `permute` → `ttnn.transpose` (10× faster)~~ this was actually just reshape-permute-reshape -> permute. Tricky to generalize in compiler. — *Attention layout*
- fuse N reductions of one matmul's output slices → 1 reduction — *Reshard + fuse collectives*
- tile-aware gather dim: pack partials into a tile dim, not the leading dim — *stat-gather repack*
- emit combine/reduce in the dtype that hits `FastReduceNC` (BF16) — *MoE tail precision*
- ~~CSE loop-invariant RoPE cos/sin broadcasts~~ — already covered (hand-added code, outside the compiled graph)

**Detail:** [MoE-compute journal](https://github.com/tenstorrent/tt-metal/blob/mvasiljevic/deepseek-decode-sharding/deepseek_codegen/graph_0/MOE_COMPUTE_JOURNAL.md) · [Decode-sharding journal](https://github.com/tenstorrent/tt-metal/blob/mvasiljevic/deepseek-decode-sharding/deepseek_codegen/graph_0/DECODE_SHARDING_JOURNAL.md) · [Latest perf profile](https://github.com/tenstorrent/tt-metal/tree/mvasiljevic/deepseek-decode-sharding/deepseek_codegen/graph_0/perf_reports/ropecse_2026_06_12_07_08_23)

**Pure CCL share** (explicit `all_gather`/`reduce_scatter`/`all_to_all`):

| part | attn | dense | moe | lm_head | **full model** |
|---|---|---|---|---|---|
| CCL % | 23% | 57% | 32% | 24% | **~29%** |

Excludes the MoE combine fused inside `MoECompute` (~35–40% incl. it). 

Note: Using trace will stabilize ccls and make them faster.

Note: Model was originally sharded the wrong way, which is out of scope of this effort. Changing it [like we are doing for glm](https://github.com/tenstorrent/tt-xla/pull/5139) would eliminate most of the ccls in the attention block.

**Reply by mvasiljevicTT** [reply]
the [#46208](https://github.com/tenstorrent/tt-metal/issues/46208) `selective_reduce_combine` L1-overlap blocker flagged above — **it's now fixed in tt-metal by [PR #46544](https://github.com/tenstorrent/tt-metal/pull/46544)** (dynamic fabric-mux buffer sizing: shrinks `num_buffers` until the mux region fits under the lowest occupied compute-L1 tensor).

That unblocked the expert/combine-concurrent restructuring (all PCC-safe, golden logits PCC held at 0.8989):

- **MoE shared-FFN CCL fusion** — `reduce_scatter + all_gather → all_gather + FastReduceNC`, which used to deterministically hang the concurrent combine, now runs clean (−87 µs/MoE layer).
- **Indexer RoPE → [`rotary_embedding_hf`](https://github.com/tenstorrent/tt-metal/tree/main/ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_hf)** — native rotate_half on half-concat, dropping the tile-crossing interleaved-pair round-trip (−53 µs/attn layer).

Now up to **~6.9 samples/sec**. Detail in the [decode-sharding journal](https://github.com/tenstorrent/tt-metal/blob/mvasiljevic/deepseek-decode-sharding/deepseek_codegen/graph_0/DECODE_SHARDING_JOURNAL.md).

**Reply by mvasiljevicTT** [reply]
> tile-aware gather dim: pack partials into a tile dim, not the leading dim — stat-gather repack

In this case that was `all_reduce` that was decomposed to rs+ag. In the workaround there is logic for choosing the gathering dim, but when it fails it falls back to dim 0. I would guess that -1 is better default value in general, we should check what is it tt-metal.

**Reply by mvasiljevicTT** [reply]
> fuse N reductions of one matmul's output slices → 1 reduction — Reshard + fuse collectives

This happens when attention weights are sharded, so QKV matmul needs to be reduced. So far, we never really wanted this scenario, it was just product of wrong sharding spec.

The pattern of op having only slices as users, which all have same users is solvable in tt-mlir by commuting slice op. SliceCommute pass was discussed a few times, and every time the conclusion was that the need for it was actually the red flag of some other issue. That is also the case here.

I guess we will add it at some point, but I don't see a real use case yet.

**Reply by mvasiljevicTT** [reply]
> emit combine/reduce in the dtype that hits FastReduceNC (BF16) — MoE tail precision

This comes from `(bf16->fp32)->sum->(fp32->bf16)` pattern in MoE tail, that should be replaced with `sum<bf16, fp32_dest_acc_en>` that will use `FastReduceNC` kernel. Part of perf boost here is probably just from removing typecasts ops. 

This pattern is currently also present in glm, and will probably be in other moe models. We should check this again once `moe_compute` lands. 

This seems like a reasonable pattern to have regardless of moe, not sure atm where are we with dtype optimizations like this?

Note: playing with dtypes is tricky when iterating on a few layer model version as it definitely lowers precision, the question is how much would it accumulate on the full model. Could this be calculated? Is it like err^n?
